### PR TITLE
improve docs for `collect` and square brackets

### LINKED ISSUE #55350: Explicit calls to type inference in array comprehensions cause equivalent expressions to have different values
I'm sure this issue is very well known among core developers, but I though it wouldn't hurt to open a Github issue to track it. If this is already covered by a different open issue, feel free to close this one.

Julia is a dynamic language, so the value of an expression should only depend on the runtime values of the components of the expression. However, this is not always the case when using generators or array comprehensions, which I think is because the implementation uses `Core.Compiler.return_type` (or similar) somewhere along the line when the input collections are empty. (Sorry, I haven't actually read through the relevant code lately, but I've looked at it in the past.)

Here's a simple example of two expressions where the components of the expression have the exact same values, but the values of the expressions are different:

Define input values:

```julia
const a = Int[]
const b = Int[]
c = Int[]
d = Int[]
g(::Vector{Int}) = 1
g(::Vector{Any}) = 2
```

Evaluate expressions:

```julia
julia> a == c, b == d
(true, true)

julia> g([x*y for x in a for y in b])
1

julia> g([x*y for x in c for y in d])
2
```

### PR DESCRIPTION
**Pull request by JeffBezanson**
fixes #55350

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -692,6 +695,21 @@ julia> collect(x^2 for x in 1:3)
...
+Collecting an empty iterator where the result type depends on type inference:
+
+```jldoctest
+julia> [rand(Bool) ? 1 : missing for _ in []]
+Union{Missing, Int64}[]
+```
+
+When the iterator is non-empty, the result type depends only on values:
+
+```
```
Comment (nsajko) [inline base/array.jl:708]:
````suggestion
```julia-repl
````

**Comment by nsajko** [comment]
Should similar notes be given for `map` and for broadcasting?

```julia-repl
julia> map(Returns(0.3), Any[])
Float64[]

julia> Returns(0.3).(Any[])
Float64[]
```

**Comment by JeffBezanson** [comment]
Yes ok, I will add links.

**Review by nsajko — APPROVED** [review]
Checking the PDF artifact shows that all doc string Markdown changes to doc strings included in the PDF are well-formed.

The `[]` doc string isn't included yet, xref #55400

**Pull request merged**